### DISCUSSION TITLE
Sometimes GitHub Actions isn't triggered by pull_request closed event

### DISCUSSION THREAD

**Discussion by suzuki-shunsuke** [community/community#26657] [Actions]
<h2>
<a name="background-1" class="anchor" href="#background-1"></a>Background</h2>
<p>We are running two GitHub Actions Workflow.</p>
<ol>
<li>workflow_dispatch event: Close stale pull requests with GitHub App</li>
<li>pull_request closed event: Create issues regarding closed pull requests with GitHub App</li>
</ol>
<p>We call GitHub API with GitHub App to close pull requests.<br>
We use this GitHub Actions to issue GitHub Token.</p>
<aside class="onebox allowlistedgeneric" data-onebox-src="https://github.com/tibdex/github-app-token">
  <header class="source">
      <img src="https://github.githubassets.com/favicons/favicon.svg" class="site-icon" width="32" height="32">

      <a href="https://github.com/tibdex/github-app-token" target="_blank" rel="noopener nofollow ugc">GitHub</a>
  </header>

  <article class="onebox-body">
    <div class="aspect-image" style="--aspect-ratio:690/345;"><img src="https://user-images.githubusercontent.com/13323303/181099090-d0df9fd4-5709-4c5c-9ff3-dc49985ba91b.png" class="thumbnail" width="690" height="345" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4c79177a90bd4ebe19a97ec17ef2831b168b8af6_2_690x345.png, https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4c79177a90bd4ebe19a97ec17ef2831b168b8af6_2_1035x517.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/4/c/4c79177a90bd4ebe19a97ec17ef2831b168b8af6.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/4/c/4c79177a90bd4ebe19a97ec17ef2831b168b8af6_2_10x10.png"></div>

<h3><a href="https://github.com/tibdex/github-app-token" target="_blank" rel="noopener nofollow ugc">GitHub - tibdex/github-app-token: Impersonate a GitHub App in a GitHub Action</a></h3>

  <p>Impersonate a GitHub App in a GitHub Action. Contribute to tibdex/github-app-token development by creating an account on GitHub.</p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

<h2>
<a name="problem-2" class="anchor" href="#problem-2"></a>Problem</h2>
<p>Sometimes GitHub Actions Workflows (2) aren’t triggered even if pull requests are closed.<br>
We think there is no problem with permission because sometimes GitHub Actions Workflows are appropriately triggered.</p>
<p>Do you have any ideas why sometimes GitHub Actions Workflows aren’t triggered expectedly?</p>
<h2>
<a name="github-actions-workflow-3" class="anchor" href="#github-actions-workflow-3"></a>GitHub Actions Workflow</h2>
<aside class="onebox githubgist" data-onebox-src="https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b">
  <header class="source">

      <a href="https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b" target="_blank" rel="noopener nofollow ugc">gist.github.com</a>
  </header>

  <article class="onebox-body">
    <h4><a href="https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b" target="_blank" rel="noopener nofollow ugc">https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b</a></h4>

  <h5>autoclose-renovate.yaml</h5>
  <pre><code class="YAML">name: Close Renovate pull requests

on:
  workflow_dispatch:
    inputs: {}
  schedule:
    # The "*" (#42, asterisk) character has special semantics in YAML, so this
    # string has to be quoted.
    - cron: '0/15 * * * *'
permissions:</code></pre>
    This file has been truncated. <a href="https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b" target="_blank" rel="noopener nofollow ugc">show original</a>
  <h5>renovate-issue-action.yaml</h5>
  <pre><code class="YAML">---
name: renovate-issue-action
on:
  pull_request:
    branches: [master]
    types:
      - closed
permissions:
  issues: write
  contents: read</code></pre>
    This file has been truncated. <a href="https://gist.github.com/suzuki-shunsuke/f2e3d52bf8f5da053e06e9d80398b27b" target="_blank" rel="noopener nofollow ugc">show original</a>

<p>
</p>

  </article>

  <div class="onebox-metadata">
    
    
  </div>

  <div style="clear: both"></div>
</aside>

**Comment by suzuki-shunsuke** [comment]
<p>When we close pull requests manually from Web UI, GitHub Actions Workflows are triggered by pull_request closed event and we can’t reproduce the problem.</p>

**Comment by jsoref** [comment]
<p>On average, this is the explanation:<br>
<a href="https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow" class="onebox" target="_blank" rel="noopener">https://docs.github.com/en/actions/using-workflows/triggering-a-workflow#triggering-a-workflow-from-a-workflow</a></p>
<p>But w/o a public repository showing the problem, it’s hard to be certain.</p>

**Comment by suzuki-shunsuke** [comment, marked as answer]
<p>I’ve asked GitHub Support this issue from <a href="https://support.github.com/" rel="noopener nofollow ugc">https://support.github.com/</a><br>
and was given the following answer.</p>
<blockquote>
<p>Hi there,</p>
<p>Thank you for your patience while the team investigated this issue.</p>
<p>I have been updated that the reason you are experiencing this issue is that GitHub doesn’t queue the merge commit job for closed PRs; So workflows using on: { pull_request: { types: [“closed”] } } will not consistently be triggered.</p>
<p>The workaround is to use a <code>pull_request_target</code> trigger as those do not require a merge commit but run the workflow from the target branch. But take note of the warnings in the docs below.</p>
<p><a href="https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target" rel="noopener nofollow ugc">https://docs.github.com/en/actions/reference/events-that-trigger-workflows#pull_request_target</a></p>
<p>The team has also suggested that you either combine both workflows or use a <code>workflow_run</code> trigger for the second workflow</p>
</blockquote>

**Reply by PColuccio** [reply]
In my case the workflow not triggering was due to a merge conflict between the pull request branch and the base branch. The `pull_request` trigger uses the merge commit between the pull request branch, and the pull request base branch. If there is a merge conflict, that merge branch cannot be created. The `pull_request_target` trigger doesn't use the merge branch, so it will run even when the branch in the pull request isn't mergable.

**Reply by nuang-ee** [reply]
@PColuccio Thank you for sharing - was exactly same on my case!
